add dtypes to range dataset op. fix #33414

ljwh opened the pull request:
Hi @jsimsa ,
This PR is just allow to specify the type of the tf.data.Dataset.range function, see issue #33414

I also add test code to range_dataset_op_test.cc and range_test.py.

jsimsa reviewed (changes requested):
@ljwh thank you very much for the change, left some minor comments. 

You will also need to run the [API compatibility test](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/api_compatibility_test.py) with `--update_goldens` flag set to True to reflect your change to the `tf.data.Dataset.range` API in the golden API files.

ljwh commented:
Hi, @jsimsa  thank you for your comments, I updated the code and **please review**.

And I have a doubt, I add output_type param this way for compatible the previous code: 
```python
def range(*args, output_type=dtypes.int64):
```
so the call must be like this:
```python
# work
range(start, stop, step, output_type=tf.int64) 
# not work
range(start, stop, step, tf.int64) 
```
I don't think define interface this way is good, do you have any suggestions?

jsimsa commented on tensorflow/python/data/kernel_tests/range_test.py:
```diff
@@ -67,6 +68,13 @@ def testStopLessThanStartWithNegativeStep(self):
     dataset = dataset_ops.Dataset.range(start, stop, step)
     self.assertDatasetProduces(dataset, expected_output=range(10, 2, -1))
 
+  def testOutputType(self):
```
my suggestion was to avoid adding a new test `testOutputType` and instead parametrize each of the above test with different output types. After my https://github.com/tensorflow/tensorflow/commit/2a2c812ab2330c9aac33335f10679a346436acfb, you can do this by replacing:

```
@combinations.generate(test_base.default_test_combinations())
def testStop(self):
  dataset = dataset_ops.Dataset.range(5)
  ...
```

with:

```
@combinations.generate(combinations.times(
    test_base.default_test_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, dtypes.float32, dtypes.float64])))
def testStop(self, output_type):
  dataset = dataset_ops.Dataset.range(5, output_type=output_type)
  ...
```

jsimsa reviewed (changes requested):
Thank you @ljwh. As for your API comment, I do not see a better alternative to the one in which `output_type` needs to be specified through a keyword that would be backwards compatible at the same time.

Also, please resolve merge conflicts.

ljwh commented:
Hi, @jsimsa , thank you for your guide, I updated the test code.

I am in doubt that why I could only pass the test code with eager mode while graph mode fail:
```python
# eager only, test passed
@combinations.generate(combinations.times(
    test_base.eager_only_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, 
        dtypes.float32, dtypes.float64, dtypes.string])))

# eager and graph, test failed
@combinations.generate(combinations.times(
    test_base.default_test_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, 
        dtypes.float32, dtypes.float64, dtypes.string])))
```

Could you please point me in the direction, thank you in advance.

jsimsa commented on tensorflow/python/data/kernel_tests/range_test.py:
```diff
@@ -18,75 +18,106 @@
+  @combinations.generate(combinations.times(
+    test_base.eager_only_combinations(),
+    combinations.combine(output_type=[dtypes.int32, dtypes.int64, 
+        dtypes.float32, dtypes.float64, dtypes.string])))
```
Does it make sense to be using `dtypes.string` for all of these examples if the start / stop / step values are integers?

What would your expected use of `tf.data.Dataset.range` with `output_type=dtypes.string` look like?

ljwh replied:
removed dtypes.string

jsimsa commented:
What is the test failure with graph mode?

ljwh commented:
Hi, @jsimsa , I found that:
```python
# next_ type defined as int64, 
# so the out_tensors is dtypes.int64
# and we can get output_dtype with   dataset()->output_dtypes()[0]
out_tensors->emplace_back(next_);
```
How to convert **next_** from **int64** to **output_dtype** here?

jsimsa commented:
You should be able to do this:

```
...
Tensor result(dataset()->output_types()[0], dataset()->output_shapes()[0]);
switch (dataset()->output_types()[0]) {
#define HANDLE_TYPE(type)                                         \
  case DataTypeToEnum<type>::value: {                             \
    result->scalar<type>()() = static_cast<type>(next_);          \
    break;                                                        \
  }
TF_CALL_NUMBER_TYPES(HANDLE_TYPE);
#undef HANDLE_TYPE
  default:
    return errors::InvalidArgument("Unsupported data type: ", DataTypeString(dataset()->output_types()[0]));
}
out_tensors->push_back(std::move(result));
...
```

ljwh commented:
Hi @jsimsa , thank you.

 I fixed the problem, please review.

gbaned commented:
@ljwh thank you for your contribution, please sign CLA.

ljwh commented:
@googlebot I fixed it.

gbaned commented:
@ljwh  Could you please check failed build errors? Thanks!

ljwh commented:
Hi, @gbaned thank you for your reminder.

Hi, @jsimsa, I updated the code for backwards compatible to CI test, **please review**.

Looks like the CI would run after pr approved.

ljwh commented:
Hi, @jsimsa 
```python
  # Support in python3 but not support in python2(Syntax error)
  def range(*args, output_type=dtypes.int64):

  # So, I changed this func and considered the backwards compatible:
  def range(*args, **kwargs):
```
Also updated some code format for pylint, **Please review**

ljwh commented:
hi @gbaned 
why ci does not run?

gbaned commented:
@ljwh  thank you for your contribution, please sign CLA.

ljwh committed:
add output_type to tf.data.RangeDataset, fix #33414.

ljwh commented:
Hi, @jsimsa 
I updated the api golden files(by run api compatibility test).
Please review.

ljwh committed:
add output_type to tf.data.RangeDataset, fix #33414.

ljwh commented:
Hi, @jsimsa 
CI build/test looks good, How to **trigger the copybara** ?

jsimsa commented:
@gbaned could you please unblock this PR? thank you

ljwh commented:
@gbaned could you please **trigger the copybara** ? thank you

gbaned commented:
@jsimsa, @ljwh There is an automatic merge process that seems to be stalled for now. Thank you!

ljwh commented:
Hi, @gbaned 
How is that going? Thank you

ljwh commented:
@jsimsa @gbaned  thank you !

jsimsa commented:
@gbaned what is this blocked on?
